gateway/mw_graphql_transport.go

```diff
@@ -81,6 +84,23 @@ func (g *GraphQLEngineTransport) handleProxyOnly(proxyOnlyCtx *GraphQLProxyOnlyC
 		return nil, err
 	}
 
+	if response.StatusCode >= http.StatusBadRequest {
+		// In proxy-only mode, we keep the upstream error message to
+		// insert into the library's error message.
+		body, err := io.ReadAll(response.Body)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			_ = response.Body.Close()
+		}()
+		// graphql-go-tools uses response.body to resolve the upstream response.
+		// It's not possible to re-use io.ReadCloser. Because of that, we keep the
+		// original error message for later use.
+		// See TT-7808
+		proxyOnlyCtx.originalResponseBody = bytes.NewReader(body)
```

Comment:
Can you consider using `nopCloserBuffer` or `nopCloser`, both exist in reverse_proxy.go, and could be isolated and moved into /internal/httputil; they allow for re-reading of an io.ReadCloser, which should match request or response bodies.